ember-core/shared/src/main/scala/org/http4s/ember/core/h2/H2Stream.scala

```diff
@@ -73,18 +73,26 @@ private[h2] class H2Stream[F[_]: Concurrent](
 
   def sendMessageBody(mess: Message[F]): F[Unit] = {
     val trailers = mess.attributes.lookup(Message.Keys.TrailerHeaders[F])
-    mess.body.chunks.noneTerminate.zipWithNext
-      .foreach {
-        case (Some(c), Some(Some(_))) =>
-          sendData(c.toByteVector, false)
-        case (Some(c), Some(None) | None) =>
-          sendData(c.toByteVector, trailers.isEmpty)
-        case (None, _) =>
-          if (trailers.isDefined) Applicative[F].unit
-          else sendData(ByteVector.empty, true)
-      }
-      .compile
-      .drain
+    val maxFrameSize = remoteSettings.map(_.maxFrameSize.frameSize)
+    maxFrameSize.flatMap(maxFrameSize =>
+      mess.body
+        .ifEmpty[F, Byte](
+          Stream.exec(
+            // Message empty with trailing headers, do nothing
+            if (trailers.isDefined) Applicative[F].unit
+            // Message empty no trailing headers, send empty bytevector
+            else sendData(ByteVector.empty, true)
```

Comment:
Not a thing, just a tiny improvement suggestion
```suggestion
            // if the Message is empty with trailing headers, do nothing
            // otherwise, if the Message is empty with no trailing headers, send an empty ByteVector
           sendData(ByteVector.empty, true).when(trailers.isEmpty)
```

Reply:
Nice! there's also `unlessA(trailers.isDefined)` , maybe that reads even better:

"send an empty bytevector unless there are trailers"

Reply:
"send an empty bytevector when there are no trailers" also works fine for me 😛

Reply:
This is fun...
Now we're always using `trailers.isEmpty` which makes me realize we only care about their absense. How about this:

```diff
   def sendMessageBody(mess: Message[F]): F[Unit] = {
-    val trailers = mess.attributes.lookup(Message.Keys.TrailerHeaders[F])
+    val noTrailers = mess.attributes.lookup(Message.Keys.TrailerHeaders[F]).isEmpty
     val maxFrameSize = remoteSettings.map(_.maxFrameSize.frameSize)
     maxFrameSize.flatMap(maxFrameSize =>
       mess.body
         .ifEmpty[F, Byte](
-          Stream.exec(
-            // send empty bytevector when there are no trailing headers
-            sendData(ByteVector.empty, true).whenA(trailers.isEmpty)
-          )
+          Stream.exec(sendData(ByteVector.empty, true).whenA(noTrailers))
         )
         .chunkLimit(maxFrameSize)
         .zipWithNext
         .foreach { case (c, nextChunk) =>
-          val isEndStream = nextChunk.isEmpty && trailers.isEmpty
+          val isEndStream = nextChunk.isEmpty && noTrailers
           sendData(c.toByteVector, isEndStream)
         }
         .compile
```
At this point the empty stream handling code is shorter than the comment explaining it.
Additionally this makes me wonder if `Vault` should have an `exists` method that returns `Boolean` :thinking:

Reply:
Mmmm, yes, I like it, I'm going with it. :)

Reply:
I decided to open a quick PR to Vault and go with `contains` instead of `exists` to match the underlying `Map` API.
https://github.com/typelevel/vault/pull/447

Reply:
Updated to use the new `contains` :)